server/src/main/java/org/elasticsearch/action/synonyms/AbstractSynonymsPagedResultAction.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.synonyms;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.synonyms.PagedResult;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.function.IntFunction;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+/**
+ * Base class for synonyms retrieval actions, including {@link GetSynonymsAction} and {@link ListSynonymsAction}.
+ * Provides basic pagination requests and responses
+ *
+ * @param <T> Type to be returned by the response, wrapped in a {@link PagedResult}
+ */
+public abstract class AbstractSynonymsPagedResultAction<T extends ActionResponse> extends ActionType<T> {
```

Comment:
Abstract class for both GET and List actions, dealing with pagination params

Reply:
Seems to be too much abstractions for simple classes like this, but that's my personal opinion. I will leave this to you, as the main developer.